harvest/storage/base.py

```diff
@@ -0,0 +1,244 @@
+                data = self.load(symbol, interval_timedelta[0], start, end)
+                if not data is None:
+                    return data 
+            return None
+
+        dt_interval = interval_to_timedelta(interval)
+
+        self.storage_lock.acquire()
+        if interval not in self.storage[symbol]:
+            # If we don't have the given interval but we a smaller one,
+            # then aggregate the data
+            intervals = [(interval, interval_to_timedelta(interval)) for interval in self.storage[symbol].keys() if interval_to_timedelta(interval) < dt_interval] 
+            if len(intervals) == 0:
+                self.storage_lock.release()
+                return None
+
+            data = self.storage[symbol][intervals[-1][0]]
+            data = self._aggregate(data, interval)
+        else:
+            data = self.storage[symbol][interval]
+        self.storage_lock.release()
+
+        # If the start and end are not defined, then set them to the 
+        # beginning and end of the data.
+        if start is None:
+            start = data.index[0]
+
+        if end is None:
+            end = data.index[-1]
+
+        return data.loc[start:end]
+
+    def data_range(self, symbol: str, interval: str) -> Tuple[dt.datetime]:
+        """
+        Returns the oldest and latest datetime of a particular symbol.
+        :symbol: a stock or crypto
+        :interval: the interval between each data point, must be atleast
+             1 minute
+        """
+        data = self.load(symbol, interval)
+        if data is None:
+            return None, None
+
+        dt_interval = interval_to_timedelta(interval)
+        return data.index[0], data.index[-1]
+
+    def _aggregate(self, data: pd.DataFrame, interval: str) -> pd.DataFrame:
+        """
+        :data: a pandas dataframe that needs to be aggregated
+        :interval: the new time between data points
+        """
+        op_dict = {
+            'open': 'first',
+            'high':'max',
+            'low':'min',
+            'close':'last',
+            'volume':'sum'
+        }
+        val, unit = utils.expand_interval(interval)
+        if unit == 'MIN':
+            val = val+'T'
+        elif unit == 'HR':
+            val = 'H'
+        else:
+            val = 'D'
+        data = data.resample(val).agg(op_dict)
+        return data
+
+
+    def _append(self, current_data: pd.DataFrame, new_data: pd.DataFrame, interval: str, remove_duplicate: bool) -> pd.DataFrame:
+        """
+        Appends the data as best it can with gaps in the data for weekends 
+        and time when no data is collected. 
+        :current_data: the current data that we have on the stock for 
+            the interval
+        :new_data: data coming from the the broker's API call
+        """
+
+        new_df = current_data.append(new_data)
+        if remove_duplicate:
+            new_df = new_df[~new_df.index.duplicated(keep='last')]
+        return new_df
+
```

Comment:
This is a crude but tried-and-tested way to implement `append`. This will be used until other parts of the code has been tested.

Reply:
The commented out append was designed to allow appending anywhere in the existing data such as inbetween or before or overlapping with the new data overriding the old one. I'll make sure to test it more.